internal/uvm/virtual_device.go

```diff
@@ -14,55 +14,104 @@ import (
 // VPCIDevice represents a vpci device. Holds its guid and a handle to the uvm it
 // belongs to.
 type VPCIDevice struct {
-	vm *UtilityVM
-	ID string
+	vm               *UtilityVM
+	VMBusGUID        string
+	deviceInstanceID string
+	refCount         uint32
 }
 
 // Release frees the resources of the corresponding vpci device
 func (vpci *VPCIDevice) Release(ctx context.Context) error {
-	if err := vpci.vm.RemoveDevice(ctx, vpci.ID); err != nil {
+	if err := vpci.vm.RemoveDeviceFromUVM(ctx, vpci); err != nil {
 		log.G(ctx).WithError(err).Warn("failed to remove VPCI device")
 		return err
 	}
 	return nil
 }
 
-// AssignDevice assigns a new vpci device to the uvm
-func (uvm *UtilityVM) AssignDevice(ctx context.Context, device hcsschema.VirtualPciDevice) (*VPCIDevice, error) {
+// AssignDeviceToUVM assigns a vpci device to the uvm
+// if the device already exists, the stored VPCIDevice's ref count is increased
+// and the VPCIDevice is returned.
+// Otherwise, a new request is made to assign the target device indicated by the deviceID
+// onto the UVM. A new VPCIDevice entry is made on the UVM and the VPCIDevice is returned
+// to the caller
+func (uvm *UtilityVM) AssignDeviceToUVM(ctx context.Context, deviceID string) (*VPCIDevice, error) {
```

Comment:
Why the name change here and for Remove? I think the `ToUVM` part is fine being implicit since it takes a `*UtilityVM` receiver.

Reply:
I asked this offline so I'll post her response. 'I just wanted to be more specific. It probably isn't necessary but I think it helps with readability since anyone reading where it is called knows immediately there is extra work going on somewhere else for assigning into the target container'

Reply:
I still would prefer the ToUVM removed.. 😁 not sure how others feel

Reply:
@katiewasnothere to me, this function appears to be doing about the same set of work as e.g. the SCSI code. We do a HCS request to map something to the UVM, and then a guest request to prepare it for use inside the UVM. I think it is okay to remove `ToUVM` here, and it is still clear that additional work is needed, since that's how most of our other code operates as well.